### ISSUE TITLE
250818 PyTorch 김광묵

### ISSUE THREAD

**Issue by JackFink** [AIBootcamp14/wiki-rider#31]
## 학습 날짜
*2025-08-18*

## 키워드 및 요약 정리
*키워드: PyTorch, 딥러닝 프레임워크, 텐서(Tensor)*

- **PyTorch**는 딥러닝 모델을 개발하는 프레임워크다.
- **핵심** 딥러닝 프레임워크 구현된 것들을 가져다 쓸수 있다.
- **요약** *맨땅 부터 코딩? PyTorch를 사용해서 아이템전으로 빠르게*

### 1. 딥러닝 프레임워크

딥러닝 모델을 프레임워크 없이 구현하려면, 모든 것을 직접 만들어야 한다.
- 모든 Neural Network Layer 구현
- Loss Function 구현
- 모든 Layer의 가중치(weight), 편향(bias)에 대한 Gradient 계산
- 경사 하강법(Gradient Descent) 등 최적화 알고리즘 구현

**딥러닝 프레임워크 구현된 것들을 가져다 쓸수 있다.**
- **제공 요소:** 모델 구성에 필요한 모든 요소를 미리 구현하여 제공
- **Loss Function:** 다양한 Loss Function 제공
- **자동 미분:** 미분 과정의 자동화
- **다양한 Optimizer:** 다양한 최적화 알고리즘을 제공한다.

### 2. 딥러닝 프레임워크 트렌드

딥러닝 프레임워크 TensorFlow, PyTorch, JAX, MANet 등

- **트렌드:** State-of-the-art(SOTA) 딥러닝 논문에서 사용되는 모델의 대부분이 PyTorch로 공개되고 있다.
- **강력한 생태계:**
    - **자연어 처리(Natural language processing):** `Hugging Face`라는 커뮤니티의 대부분의 공개 모델이 PyTorch.
    - **컴퓨터 비전(Computer vision):** `timm` `segmentation_models_pytorch` 라이브러리 등 과 같은 라이브러리를 사전 학습된 모델을 편리하게 학습에 활용할 수 있다.
    - **대규모 언어 모델(Large language model):** 최신 LLM들 또한 PyTorch 프레임 워크로 공개되고 있다.

### 3. 환경설정
- Anaconda (아나콘다)
    - [홈페이지](https://www.anaconda.com/) 운영체제에 맞게 설치를 진행
    - (base) 표시가 나타나면 정상
- NVIDIA GPU가 없는 경우
    - 설치해둔 가상환경에서 PyTorch를 ‘CPU-only’ 버전으로 설치할 수 있다.
        ```cmd
            >>>import torch
            입력 후 아무것도 출력되지 않는다면 정상이다.
            무언가 나온다면 꽝이다
        ```
- NVIDIA GPU가 있는 경우
    - 설치 
        ```cmd
            ~$ nvidia-smi (GPU 정보 확인)
            ~$ nvidia-smi -L (GPU 이름 확인)
        ```
    - https://en.wikipedia.org/wiki/CUDA  GPU에 맞는 ‘Compute capability(version)’를 통해 ‘CUDA SDK(version)’로 현재 GPU에 맞는 CUDA 버전을 찾아 확인한다.    

    - 확인한 버전을 https://pytorch.org/get-started/previous-versions/ 를 참고해서 진행한다.

        ```cmd
            ~$ conda install pytorch==2.0.0 torchvision==0.15.0 torchaudio==2.5.1  pytorch-cuda=11.8 -c pytorch -c nvidia
        ```

## ❓ 어려웠던 부분
- 

## 질문 또는 토론 유도
- *TensorFlow는 많이 들어봤는데, PyTorch는 낯설다.*
- *PyTorch가 논문 분야에 특화된 것인가?*
- *AMD 버려?*

## 참고 자료 (선택)
-온라인 강의

**Comment by parkjunohjjang12** [comment]
안녕하세요 김광묵 님! 
키워드부터 개념 정리까지 체계적이고 재미있게 적어주셔서 잘 읽었습니다!
질문해 주신 부분에 대해서 피드백 드리도록 하겠습니다.

---

1️⃣ 질문 답변

**TensorFlow는 많이 들어봤는데, PyTorch는 낯설다.**
- TensorFlow는 과거에 표준처럼 사용되었으나, PyTorch가 TensorFlow 대비 개발 난이도 낮고 디버깅의 용이성으로 인해 PyTorch의 사용 비중이 늘어나게 되었습니다.
- 이로인해, 많은 엔지니어들과 연구자들이 구현하기에는 PyTorch 사용이 용이하기 때문에 대부분의 사람들이 PyTorch를 많이 사용하고 있으며, DeepMind의 경우에는 JAX를 많이 이용하는 편으로 알고 있습니다!

**PyTorch가 논문 분야에 특화된 것인가?**
- 앞서 말씀드린 것처럼 결국 구현의 용이성과 가독성 때문에 대부분은 PyTorch를 사용하고 있습니다.
- 저같은 경우에도 논문을 위한 구현 뿐 아니라, 프로젝트 등에 있어서도 다른 분들에게 공유를 드려야하기 때문에 통일성의 목적에서도 PyTorch를 많이 사용하는 것 같습니다!

**AMD?**
- PyTorch는 NVIDIA  GPU를 중심으로 최적화되어 있기 때문에, AMD보다는 NVIDIA GPU를 많이 쓰는 편이긴 합니다.
- PyTorch에서 AMD도 지원하는 것으로 알고 있으나, CUDA만큼 생태계는 크지 않다는 점이 존재합니다.
- 이로인해 대부분 논문 구현이나 학습에 있어서 NVIDIA가 훨씬 유리하지만, AMD도 지원을 하기 때문에 무조건 이에 대한 선택지는 존재할 것으로 보입니다!

---
2️⃣ Tip
- PyTorch 설치 시 가장 헷갈리는 부분이 CUDA 버전입니다.
- nvidia-smi로 드라이버 버전을 확인 → PyTorch 공식 페이지에서 맞는 CUDA 버전을 찾아 설치하는 방식이 가장 안전하니 이점 확인하시는 것이 좋을 것 같습니다!!

---

정리하자면, PyTorch는 연구 분야 표준이고, 실무에서도 대부분 사용하는 것으로 알고 있으시면 될 것 같습니다. AMD도 완전히 버려진 건 아니지만, 현 시점에서는 NVIDIA 생태계가 훨씬 강력합니다. 

감사합니다!😀
